File: app/laravel/src/Product/Application/Queries/GetProduct/GetProductHandler.php
<?php

declare(strict_types=1);

namespace Src\Product\Application\Queries\GetProduct;

use Src\Shared\Application\Bus\QueryHandler;
use Src\Shared\Domain\Exceptions\DomainException;
use Src\Product\Application\DTOs\ProductDTO;
use Src\Product\Domain\Repositories\ProductRepositoryInterface;
use Src\Product\Domain\ValueObjects\ProductId;

/**
 * GetProductHandler - Tək məhsul sorğusunu icra edən handler.
 *
 * QueryHandler Command-dan fərqli olaraq nəticə qaytarır.
 * Bu handler ProductDTO qaytarır - Domain Entity-ni birbaşa qaytarmırıq.
 *
 * Niyə Entity əvəzinə DTO qaytarırıq?
 * - Entity-nin daxili strukturunu gizləyirik.
 * - API cavabını Entity-dən asılı olmadan formatlaya bilərik.
 * - Entity-nin metodlarına xaricdən müraciət edilməsinin qarşısını alırıq.
 */
final class GetProductHandler implements QueryHandler
{
    public function __construct(
        private readonly ProductRepositoryInterface $repository,
    ) {
    }

    /**
     * @param GetProductQuery $query Məhsul sorğusu
     * @return ProductDTO Tapılan məhsulun DTO-su
     * @throws DomainException Əgər məhsul tapılmadıqda
     */
    public function handle(GetProductQuery $query): ProductDTO
    {
        $product = $this->repository->findById(
            ProductId::fromString($query->productId)
        );

        if ($product === null) {
            throw new DomainException(
                "Məhsul tapılmadı. ID: {$query->productId}"
            );
        }

        // Entity-dən DTO-ya çeviririk
        return ProductDTO::fromEntity($product);
    }
}

File: app/spring/src/main/java/az/ecommerce/product/application/query/getproduct/GetProductHandler.java
package az.ecommerce.product.application.query.getproduct;

import az.ecommerce.product.application.dto.ProductDto;
import az.ecommerce.product.domain.repository.ProductRepository;
import az.ecommerce.product.domain.valueobject.ProductId;
import az.ecommerce.shared.application.bus.QueryHandler;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import org.springframework.stereotype.Service;

@Service
public class GetProductHandler implements QueryHandler<GetProductQuery, ProductDto> {

    private final ProductRepository repository;

    public GetProductHandler(ProductRepository repository) { this.repository = repository; }

    @Override
    public ProductDto handle(GetProductQuery query) {
        return repository.findById(new ProductId(query.productId()))
                .map(ProductDto::fromDomain)
                .orElseThrow(() -> new EntityNotFoundException("Product", query.productId().toString()));
    }
}
